Agent: Include Cilium version in output of 'cilium status --verbose'

romanspb80 opened the pull request:
The status output doesn't contain the version of the Daemon agent and shows double 'OK' for state:
      $ cilium status --verbose
      ...
      Cilium:                 Ok        OK
      ...
This commit fix double 'OK' and add the version to output:
      ...
      Cilium:                 Ok        Ver: 1.9.90   Rev: 6d3143fd0
      ...
    
Fixes: #14154
    
Signed-off-by: Roman Ptitcyn romanspb@yahoo.com

sayboras commented on daemon/cmd/status.go:
```diff
@@ -599,16 +600,22 @@ func (d *Daemon) getStatus(brief bool) models.StatusResponse {
 		sr.Cilium = &models.Status{
 			State: models.StatusStateWarning,
-			Msg:   "Stale status data",
+			Msg:   msg + ciliumVer,
```
while string concatenation is faster, it's hard to read and error prone (e.g. ciliumVer _must_ have space prefix, msg should be ending with dot or comma, etc). I would prefer to go with fmt.Sprintf() in general.

romanspb80 replied:
> Thanks for PR.
> 
> Based on my understanding, version will be printed in case of any failed check. This could be
> redudant in some cases. What do you think about always printing version details at the beginning ?

Thanks Tam for the suggestion! Printing version details at the beginning is a good idea. But there are some entities which printing status at the beginning. For example:
$ cilium status
KVStore:                Ok   etcd: 1/1 connected, lease-ID=29c6767063b65989, lock lease-ID=29c6767063b6598b, has-quorum=true: https://192.168.33.11:2379 - 3.4.13 (Leader)
Kubernetes:             Ok   1.20+ (v1.20.0-rc.0) [linux/amd64]
Cilium:                 Ok            Ver: 1.9.90  Rev: 25d8615be
Cilium health daemon:   Ok   
BandwidthManager:       Disabled
Host Routing:           Legacy
Proxy Status:           OK, ip 10.11.0.128, 0 redirects active on ports 10000-20000
Hubble:                 Ok              Current/Max Flows: 471/4095 (11.50%), Flows/s: 11.86   Metrics: Disabled

Maybe we will return to this idea, but a little later, after feedback? Because perhaps it would be more correct to make the output uniform for all entities.

romanspb80 replied:
> while string concatenation is faster, it's hard to read and error prone (e.g. ciliumVer _must_ have space prefix, msg should be ending with dot or comma, etc). I would prefer to go with fmt.Sprintf() in general.

OK. I've implemented fmt.Sprintf() in the next commit. And regarding the question about the order of output I swapped msg and ciliumVer. Could you please check:
$ cilium status
Cilium:                 Failure   [Ver: 1.9.90  Rev: 25d8615be]    Kvstore service is not ready

sayboras reviewed (commented):
Thanks for PR. 

Based on my understanding, version will be printed in case of any failed check. This could be 
 redudant in some cases. What do you think about always printing version details at the beginning ?

aanm reviewed (changes requested):
The changes look good overall but there are a couple things I don't understand:
1) Why are 4 spaces between `ciliumVer` and `msg`?
2) What's the reasoning to always print the cilium version in a failure scenario instead of only printing the Cilium versions int the Cilium-specific line?

christarazi commented on daemon/cmd/status.go:
```diff
@@ -599,16 +600,22 @@ func (d *Daemon) getStatus(brief bool) models.StatusResponse {
 	sr.Stale = stale
 
+	//CiliumVersion definition
+	Ver := v.GetCiliumVersion()
```
```suggestion
	ver := v.GetCiliumVersion()
```

christarazi commented on daemon/cmd/status.go:
```diff
@@ -46,6 +46,7 @@ import (
 	"github.com/cilium/cilium/pkg/option"
 	"github.com/cilium/cilium/pkg/rand"
 	"github.com/cilium/cilium/pkg/status"
+	v "github.com/cilium/cilium/pkg/version"
```
Nit: is there a reason to do a named import here? Why not just use the package name as long as there's no conflicts already, which there shouldn't be.

romanspb80 replied:
> Nit: is there a reason to do a named import here? Why not just use the package name as long as there's no conflicts already, which there shouldn't be.

Thanks for the remark.
Done. Please check

christarazi reviewed (changes requested):
A few nits below. LGTM besides what others have suggested. Also, please squash all the commits into one when you think it's ready to merge.

romanspb80 committed:
Agent: Include Cilium version in output of cilium status --verbose

The output doesn't contain the version of the Daemon agent and shows double 'OK' for status:
    $ cilium status --verbose
    ...
    Cilium:                 Ok        OK
    ...
This commit fix double 'OK', add the cilium version and delete unnecessary space in the output of the cilium status:
    ...
Cilium:                 Ok   1.9.90 (v.1.9.90-r.d3428c13e)
    ...

Fixes: #14154

Signed-off-by: Roman Ptitcyn <romanspb@yahoo.com>

aanm commented on daemon/cmd/status.go:
```diff
@@ -599,16 +600,22 @@ func (d *Daemon) getStatus(brief bool) models.StatusResponse {
 		sr.Cilium = &models.Status{
 			State: models.StatusStateWarning,
-			Msg:   "Stale status data",
+			Msg:   fmt.Sprintf("%s    %s", ciliumVer, msg),
```
it seems the space was not removed? As well as in the remaining lines.

romanspb80 replied:
> it seems the space was not removed? As well as in the remaining lines.

Yes, it is. This is OK. You only mentioned the space between State and Version: 

> The space between the Ok and the [Ver: 1.9.90 Rev: 7ce4c3609] seems unnecessary. Also, to make it consistent maybe we should make it between parenthesis, similar to what is done for Kubernetes, and not square brackets.

And this is a fix in accordance with your remark:
`
pkg/client/client.go

	if sr.Cilium != nil {
		-fmt.Fprintf(w, "Cilium:\t%s\t%s\n", sr.Cilium.State, sr.Cilium.Msg)
		+fmt.Fprintf(w, "Cilium:\t%s   %s\n", sr.Cilium.State, sr.Cilium.Msg)
	}
`

But the space between Version and Msg is needed as a delimeter. Please, see comment from [sayboras](https://github.com/sayboras):

> while string concatenation is faster, it's hard to read and error prone (e.g. ciliumVer must have space prefix, msg should be ending with dot or comma, etc). I would prefer to go with fmt.Sprintf() in general.

sayboras reviewed (approved):
LGTM 💯

fristonio reviewed (approved):
LGTM :rocket:

jibi reviewed (approved):
Looks good :+1:
